server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java

```diff
@@ -83,57 +92,80 @@ public IFill copy() {
   }
 
   @Override
-  Filter constructFilter() {
+  void constructFilter() {
     Filter lowerBound = beforeRange == -1 ? TimeFilter.gtEq(Long.MIN_VALUE)
         : TimeFilter.gtEq(queryTime - beforeRange);
     Filter upperBound = afterRange == -1 ? TimeFilter.ltEq(Long.MAX_VALUE)
         : TimeFilter.ltEq(queryTime + afterRange);
     // [queryTIme - beforeRange, queryTime + afterRange]
-    return FilterFactory.and(lowerBound, upperBound);
+    beforeFilter = FilterFactory.and(lowerBound, TimeFilter.ltEq(queryTime));
+    afterFilter = FilterFactory.and(TimeFilter.gtEq(queryTime), upperBound);
   }
 
   @Override
-  public void configureFill(Path path, TSDataType dataType, long queryTime,
-      Set<String> sensors, QueryContext context)
-      throws StorageEngineException, QueryProcessException {
+  public void configureFill(
+      Path path, TSDataType dataType, long queryTime, Set<String> sensors, QueryContext context) {
+    this.seriesPath = path;
     this.dataType = dataType;
     this.queryTime = queryTime;
-    Filter timeFilter = constructFilter();
-    dataReader = new SeriesRawDataBatchReader(path, sensors, dataType, context,
-        QueryResourceManager.getInstance().getQueryDataSource(path, context, timeFilter),
-        timeFilter, null, null);
+    this.context = context;
+    this.allSensors = sensors;
+    constructFilter();
   }
 
   @Override
-  public TimeValuePair getFillResult() throws IOException, UnSupportedFillTypeException {
-    TimeValuePair beforePair = null;
-    TimeValuePair afterPair = null;
-    while (batchData.hasCurrent() || dataReader.hasNextBatch()) {
-      if (!batchData.hasCurrent() && dataReader.hasNextBatch()) {
-        batchData = dataReader.nextBatch();
-      }
-      afterPair = new TimeValuePair(batchData.currentTime(), batchData.currentTsPrimitiveType());
-      batchData.next();
-      if (afterPair.getTimestamp() <= queryTime) {
-        beforePair = afterPair;
-      } else {
-        break;
-      }
-    }
+  public TimeValuePair getFillResult()
+      throws IOException, QueryProcessException, StorageEngineException {
+    QueryDataSource dataSource =
+        QueryResourceManager.getInstance().getQueryDataSource(seriesPath, context, beforeFilter);
+    LastPointReader lastReader =
+        new LastPointReader(seriesPath, dataType, allSensors, context, dataSource, queryTime, beforeFilter);
+
+    TimeValuePair beforePair = lastReader.readLastPoint();
+    TimeValuePair afterPair = calculatFirstPointAfterQueryTime();
 
     // no before data or has data on the query timestamp
-    if (beforePair == null || beforePair.getTimestamp() == queryTime) {
+    if (beforePair.getValue() == null || beforePair.getTimestamp() == queryTime) {
```

Comment:
add the LinearFill in UserGuide, explain how we fill when beforePair or afterPair is null